sentry-opentelemetry/lib/sentry/opentelemetry/span_processor.rb

```diff
@@ -0,0 +1,159 @@
+    TraceData = Struct.new(:trace_id, :span_id, :parent_span_id, :parent_sampled, :baggage)
+
+    class SpanProcessor < ::OpenTelemetry::SDK::Trace::SpanProcessor
+      include Singleton
+
+      SEMANTIC_CONVENTIONS = ::OpenTelemetry::SemanticConventions::Trace
+
+      # The mapping from otel span ids to sentry spans
+      # @return [Hash]
+      attr_reader :span_map
+
+      def initialize
+        @span_map = {}
+      end
+
+      def on_start(otel_span, parent_context)
+        return unless Sentry.initialized? && Sentry.configuration.instrumenter == :otel
+        return unless otel_span.context.valid?
+        return if from_sentry_sdk?(otel_span)
+
+        trace_data = get_trace_data(otel_span, parent_context)
+
+        sentry_parent_span = @span_map[trace_data.parent_span_id] if trace_data.parent_span_id
+
+        sentry_span = if sentry_parent_span
+          Sentry.configuration.logger.info("Continuing otel span #{otel_span.name} on parent #{sentry_parent_span.op}")
+
+          sentry_parent_span.start_child(
+            span_id: trace_data.span_id,
+            description: otel_span.name,
+            start_timestamp: otel_span.start_timestamp / 1e9
+          )
+        else
+          Sentry.configuration.logger.info("Starting otel transaction #{otel_span.name}")
+
+          options = {
+            instrumenter: :otel,
+            name: otel_span.name,
+            span_id: trace_data.span_id,
+            trace_id: trace_data.trace_id,
+            parent_span_id: trace_data.parent_span_id,
+            parent_sampled: trace_data.parent_sampled,
+            baggage: trace_data.baggage,
+            start_timestamp: otel_span.start_timestamp / 1e9
+          }
+
+          Sentry.start_transaction(**options)
+        end
+
+        @span_map[trace_data.span_id] = sentry_span
+      end
+
+      def on_finish(otel_span)
+        return unless Sentry.initialized? && Sentry.configuration.instrumenter == :otel
+        return unless otel_span.context.valid?
+
+        sentry_span = @span_map.delete(otel_span.context.hex_span_id)
+        return unless sentry_span
+
+        sentry_span.set_op(otel_span.name)
+
+        if sentry_span.is_a?(Sentry::Transaction)
+          sentry_span.set_name(otel_span.name)
+          sentry_span.set_context(:otel, otel_context_hash(otel_span))
+        else
+          update_span_with_otel_data(sentry_span, otel_span)
+        end
+
+        Sentry.configuration.logger.info("Finishing sentry_span #{sentry_span.op}")
+        sentry_span.finish(end_timestamp: otel_span.end_timestamp / 1e9)
+      end
+
+      def clear
+        @span_map = {}
+      end
+
+      private
+
+      def from_sentry_sdk?(otel_span)
+        dsn = Sentry.configuration.dsn
+        return false unless dsn
+
+        if otel_span.name.start_with?("HTTP")
+          # only check client requests, connects are sometimes internal
+          return false unless %i(client internal).include?(otel_span.kind)
```

Comment:
`%i(client internal)` could be a constant so we don't need to init a new array repeatedly.